### PR TITLE
Refactor usage status tests #335

### LINKED ISSUE #335: Refactor usage status repo, service and e2e tests
Look through the new refactored tests (see Systems, Manufacturers and Catalogue Categories) and refactor the unit and e2e tests for usage statuses to match.

### PR DESCRIPTION
**Pull request by MatteoGuarnaccia5**
## Description

see #335

## Testing instructions

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage

## Agile board tracking
closes #335

### CODE REVIEW

File: test/e2e/test_usage_status.py [hunk 1]
```diff
@@ -2,174 +2,261 @@
...
-    """Test creating a usage status"""
-    response = test_client.post("/v1/usage-statuses", json=USAGE_STATUS_POST_A)
+class CreateDSL:
+    """Base class for create tests."""
 
-    assert response.status_code == 201
-    assert response.json() == USAGE_STATUS_POST_A_EXPECTED
+    test_client: TestClient
 
+    _post_response: Response
```
Comment (joelvdavies) [inline test/e2e/test_usage_status.py:30]:
```suggestion
    _post_response_usage_status: Response
```

File: test/e2e/test_usage_status.py [hunk 2]
```diff
@@ -2,174 +2,261 @@
...
+        self.check_post_usage_status_failed_with_detail(409, "A usage status with the same value already exists")
 
-    response = test_client.get(f"/v1/usage-statuses/{response.json()['id']}")
 
-    assert response.status_code == 200
-    assert response.json() == USAGE_STATUS_POST_A_EXPECTED
+class GetDSL(CreateDSL):
+    """Base class for get tests"""
 
+    _get_response = Response
```
Comment (joelvdavies) [inline test/e2e/test_usage_status.py:89]:
```suggestion
    _get_response_usage_status: Response
```

File: test/e2e/test_usage_status.py [hunk 3]
```diff
@@ -2,174 +2,261 @@
...
-    response = test_client.post("/v1/catalogue-categories", json=CATALOGUE_CATEGORY_POST_A)
-    catalogue_category = response.json()
 
-    response = test_client.post("/v1/systems", json=SYSTEM_POST_A)
-    system_id = response.json()["id"]
+class DeleteDSL(ListDSL):
+    """Base class for delete tests."""
 
-    response = test_client.post("/v1/manufacturers", json=MANUFACTURER_POST)
+    _delete_response: Response
```
Comment (joelvdavies) [inline test/e2e/test_usage_status.py:177]:
```suggestion
    _delete_response_usage_status: Response
```

File: test/e2e/test_usage_status.py [hunk 4]
```diff
@@ -2,174 +2,261 @@
 End-to-End tests for the Usage status router
```
Comment (joelvdavies) [inline test/e2e/test_usage_status.py:2]:
```suggestion
End-to-End tests for the usage status router
```

File: test/e2e/test_usage_status.py [hunk 5]
```diff
@@ -2,174 +2,261 @@
...
-    test_client.post("/v1/usage-statuses", json=USAGE_STATUS_POST_A)
-    response = test_client.post("/v1/usage-statuses", json=USAGE_STATUS_POST_A)
+    def post_usage_status(self, usage_status_post_data: dict) -> Optional[str]:
+        """
+        Posts a usage status with the given data, returns the ID of the created usage status if successful.
 
-    assert response.status_code == 409
-    assert response.json()["detail"] == "A usage status with the same value already exists"
+        :param usage_status_post_data: Dictionary containing the usage status data as would be required for a 
+        `UsageStatusPostSchema`.
```
Comment (joelvdavies) [inline test/e2e/test_usage_status.py:42]:
```suggestion
        :param usage_status_post_data: Dictionary containing the usage status data as would be required for a 
                                                            `UsageStatusPostSchema`.
```

File: test/e2e/test_usage_status.py [hunk 6]
```diff
@@ -2,174 +2,261 @@
...
-def test_get_usage_statuses(test_client):
-    """
-    Test getting a list of usage statuses
-    """
-    test_client.post("/v1/usage-statuses", json=USAGE_STATUS_POST_A)
-    test_client.post("/v1/usage-statuses", json=USAGE_STATUS_POST_B)
-    test_client.post("/v1/usage-statuses", json=USAGE_STATUS_POST_C)
-    test_client.post("/v1/usage-statuses", json=USAGE_STATUS_POST_D)
+        :param expected_usage_status_get_data: Dictionary containing the expected usage status data as would be required 
+        for a `UsageStatusSchema`.
```
Comment (joelvdavies) [inline test/e2e/test_usage_status.py:53]:
```suggestion
        :param expected_usage_status_get_data: Dictionary containing the expected usage status data as would be required 
                                                                          for a `UsageStatusSchema`.
```

File: test/e2e/test_usage_status.py [hunk 7]
```diff
@@ -2,174 +2,261 @@
...
+    def test_create_usage_status(self):
+        """Test creating a usage status"""
 
-    assert response.status_code == 200
-    assert response.json() == []
+        self.post_usage_status(USAGE_STATUS_POST_DATA_NEW)
+        self.check_post_usage_status_success(USAGE_STATUS_GET_DATA_NEW)
 
+    def test_create_usage_status_with_duplicate_name(self):
+        """Test creating a usage status with a duplicate value"""
```
Comment (joelvdavies) [inline test/e2e/test_usage_status.py:79]:
```suggestion
    def test_create_usage_status_with_duplicate_value(self):
        """Test creating a usage status with a duplicate value."""
```

File: test/e2e/test_usage_status.py [hunk 8]
```diff
@@ -2,174 +2,261 @@
...
+        self._get_response = self.test_client.get(f"/v1/usage-statuses/{usage_status_id}")
 
-    assert response.status_code == 404
-    assert response.json()["detail"] == "Usage status not found"
+    def check_get_usage_status_success(self, expected_usage_status_get_data: dict) -> None:
+        """
+        Checks that a prior call to `get_usage_status` gave a successful response with the expected data returned.
 
+        :param expected_usage_status_get_data: Dictionary containing the expected usage status data as would be required 
+        for a `UsageStatusSchema`.
```
Comment (joelvdavies) [inline test/e2e/test_usage_status.py:104]:
```suggestion
        :param expected_usage_status_get_data: Dictionary containing the expected usage status data as would be required 
                                                                           for a `UsageStatusSchema`.
```

File: test/e2e/test_usage_status.py [hunk 9]
```diff
@@ -2,174 +2,261 @@
...
+        self.check_get_usage_status_failed_with_detail(404, "Usage status not found")
 
 
-def test_delete_with_an_invalid_id(test_client):
-    """Test trying to delete a usage_status with an invalid ID"""
+class ListDSL(GetDSL):
+    """Base class for list tests."""
 
-    response = test_client.delete("/v1/usage-statuses/invalid")
+    def get_usage_statuss(self) -> None:
```
Comment (joelvdavies) [inline test/e2e/test_usage_status.py:143]:
```suggestion
    def get_usage_status(self) -> None:
```

File: test/mock_data.py [hunk 10]
```diff
@@ -21,6 +21,36 @@
...
+    "code": "new",
+    "id": ANY,
+}
+
+USAGE_STATUS_GET_DATA_USED = {
+    **USAGE_STATUS_POST_DATA_USED,
+    **CREATED_MODIFIED_GET_DATA_EXPECTED,
+    "code": "used",
+    "id": ANY,
+}
```
Comment (joelvdavies) [inline test/mock_data.py:52]:
```suggestion

# New

USAGE_STATUS_POST_DATA_NEW = {"value": "New"}

USAGE_STATUS_IN_DATA_NEW = {
    **USAGE_STATUS_POST_DATA_NEW,
    "code": "new",
}

USAGE_STATUS_GET_DATA_NEW = {
    **USAGE_STATUS_POST_DATA_NEW,
    **CREATED_MODIFIED_GET_DATA_EXPECTED,
    "code": "new",
    "id": ANY,
}

# Used

USAGE_STATUS_POST_DATA_USED = {"value": "Used"}

USAGE_STATUS_IN_DATA_USED = {
    **USAGE_STATUS_POST_DATA_USED,
    "code": "used",
}

USAGE_STATUS_GET_DATA_USED = {
    **USAGE_STATUS_POST_DATA_USED,
    **CREATED_MODIFIED_GET_DATA_EXPECTED,
    "code": "used",
    "id": ANY,
}
```

File: test/unit/repositories/test_usage_status.py [hunk 11]
```diff
@@ -17,235 +22,421 @@
...
+
+        self.mock_session = MagicMock()
+        yield
+
+    def mock_is_duplicate_usage_status(self, duplicate_usage_status_in_data: Optional[dict] = None) -> None:
+        """
+        Mocks database methods appropriately for when the `_is_duplicate_usage_status` repo method will be called.
+
+        :param duplicate_usage_status_in_data: Either `None` or a dictionary containing usage status data for a 
+        duplicate usage status.
```
Comment (joelvdavies) [inline test/unit/repositories/test_usage_status.py:54]:
```suggestion
        :param duplicate_usage_status_in_data: Either `None` or a dictionary containing usage status data for a 
                                                                        duplicate usage status.
```

File: test/unit/repositories/test_usage_status.py [hunk 12]
```diff
@@ -17,235 +22,421 @@
...
-    with pytest.raises(MissingRecordError) as exc:
-        usage_status_repository.delete(usage_status_id)
-    assert str(exc.value) == f"No usage status found with ID: {usage_status_id}"
-    database_mock.usage_statuses.delete_one.assert_called_once_with(
-        {"_id": CustomObjectId(usage_status_id)}, session=None
-    )
+class TestList(ListDSL):
+    """Tests for listing usage statuses."""
 
+    # pylint: disable=duplicate-code
```
Comment (joelvdavies) [inline test/unit/repositories/test_usage_status.py:303]:
```suggestion
```
We have been putting
```
# Expect some duplicate code inside tests as the tests for the different entities can be very similar
# pylint: disable=duplicate-code
```
at the top of our test files instead. Could you also remove the corresponding enable below.

**Review by joelvdavies — CHANGES_REQUESTED** [review]

File: test/e2e/test_usage_status.py [hunk 13]
```diff
@@ -1,181 +1,264 @@
...
-            ],
-        ),
-    }
+        manufacturer_id = response.json()["id"]
 
-    response = test_client.post("/v1/catalogue-items", json=catalogue_item_post)
-    catalogue_item_id = response.json()["id"]
+        catalogue_item_post = {
+            **CATALOGUE_ITEM_POST_A,
+            "properties": [],
```
Comment (VKTB) [inline test/e2e/test_usage_status.py:235]: We can probably just use `CATALOGUE_ITEM_DATA_REQUIRED_VALUES_ONLY`?
```suggestion
            **CATALOGUE_ITEM_DATA_REQUIRED_VALUES_ONLY,
```

File: test/e2e/test_usage_status.py [hunk 14]
```diff
@@ -1,181 +1,264 @@
...
-        **ITEM_DATA_REQUIRED_VALUES_ONLY,
-        "catalogue_item_id": catalogue_item_id,
-        "system_id": system_id,
-        "usage_status_id": usage_status_id,
-    }
-    # pylint: enable=duplicate-code
-    response = test_client.post("/v1/items", json=item_post)
+        item_post = {
+            **ITEM_POST,
+            "properties": [],
```
Comment (VKTB) [inline test/e2e/test_usage_status.py:245]: We can probably just use `ITEM_DATA_REQUIRED_VALUES_ONLY`?
```suggestion
            **ITEM_DATA_REQUIRED_VALUES_ONLY,
```

**Review by VKTB — CHANGES_REQUESTED** [review]

File: test/e2e/test_manufacturer.py [hunk 15]
```diff
@@ -66,17 +66,6 @@ def check_post_manufacturer_failed_with_detail(self, status_code: int, detail: s
...
-    def check_post_manufacturer_failed_with_validation_message(self, status_code: int, message: str) -> None:
-        """
-        Checks that a prior call to `post_manufacturers` gave a failed response with the expected code and pydantic
-        validation error message.
-
-        :param status_code: Expected status code to be returned.
-        :param message: Expected pydantic validation error message to be returned.
-        """
-        assert self._post_response_manufacturer.status_code == status_code
-        assert self._post_response_manufacturer.json()["detail"][0]["msg"] == message
```
Comment (VKTB) [inline test/e2e/test_manufacturer.py:78]: Unused

**Comment by VKTB** [comment]
@joelvdavies Could you please have a look at the changes I requested and review the code I pushed to address them? I am not sure why codecov is reporting a drop of 22.35% in the coverage..

**Review by joelvdavies — APPROVED** [review]

**Review by VKTB — APPROVED** [review]

**Pull request merged**